### PR TITLE
Server module

### PR DESCRIPTION
**Pull request by grilledham**
### Send messages to discord
```lua
local Server = require 'features.server'
Server.to_discord('Hello from scenario script!')
```
There are other methods for sending messages to discord, but this is the recommended way as it provides markdown sanitation server side. 
Also the bot is rate limited so no spam please.

### Persistent Scenario Data
All data can be viewed live at http://redmew.com/admin/scenariodata
Feel free to make your own data_sets for testing or for actual use.

#### Send data to web storage
```lua
local Server = require 'features.server'
Server.set_data('my data set', 'key 1', 123)
Server.set_data('my data set', 'key 2', 'abc')
Server.set_data('my data set', 'key 3', {'some', 'data', ['is_set'] = true})

Server.set_data('my data set', 'key 1', nil) -- this will remove 'key 1'
Server.set_data('my data set', 'key 2', 'def') -- this will change the value for 'key 2' to 'def'
```
All data types except functions can be stored in web storage. key and data_set parameters have to be strings.

#### Get data from web storage
```lua
local Server = require 'features.server'
local Token = require 'utils.global_token'

local callback =
    Token.register(
    function(data)
        local data_set = data.data_set
        local key = data.key
        local value = data.value -- will be nil if no data

        game.print(data_set .. ':' .. key .. ':' .. tostring(value))
    end
)

Server.try_get_data('my data set', 'key 1', callback)
```
#### Get all data for data_set from web storage
```lua
local Server = require 'features.server'
local Token = require 'utils.global_token'

local callback =
    Token.register(
    function(data)
        local data_set = data.data_set
        local entries = data.entries -- will be nil if no data
        local value2 = entries['key 2']
        local value3 = entries['key 3']
    end
)

Server.try_get_all_data('my data set', callback)
```
#### Track changes made to the data from other servers or from web
```lua
local Server = require 'features.server'
Server.on_data_set_changed(
    'my data set',
    function(data)
        local data_set = data.data_set
        local key = data.key
        local value = data.value -- will be nil if data was removed.
    end
)
```
To prevent desyncs `Server.on_data_set_changed` should be called before on_init or on_load is called, just like events

#### Event for when the server is started.

the on_load event is a poor choice fro fetching data from the web server as it runs for each player that joins. on_init is also a poor choice as it will prevent you from making saves in single player and then loading them onto the server.

So I added a `Server.events.on_server_started` event
```lua
local Event = require 'utils.event'
local Server = require 'features.server'

Event.add(
    (Server.events.on_server_started),
    function()
        Server.try_get_all_data('regulars', callback)
    end
)
```

### Start a Scenario
```lua
local Server = require 'features.server'
Server.start_scenario('my_scenario_name')
```
This stops and saves the current game, then loads the scenario.
The name has to match the name of the scenario in the scenarios table on http://redmew.com/admin

### Other changes
I also changed the UserGroups module to take advantage of the web storage for regulars and donators. With this PR we should never have to manually update the regulars between sessions

### CODE REVIEW

File: server.lua [hunk 1]
```diff
@@ -0,0 +1,294 @@
+local Token = require 'utils.global_token'
+
+local Public = {}
+
+local raw_print = print
+function print(str)
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server.lua]: print isn't read-only

File: server.lua [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+local data_set_tag = '[DATA-SET]'
+local data_get_tag = '[DATA-GET]'
+local data_get_all_tag = '[DATA-GET-ALL]'
+local data_tracked_tag = '[DATA-TRACKED]'
+
+Public.raw_print = raw_print
+
+local data_set_handlers = {}
+
+defines.events.on_server_started = script.generate_event_name()
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server.lua]: wat?

File: server.lua [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+--- The event id for the on_server_started event.
+-- The event is raised whenever the server goes from the starting state to the running state.
+-- It provides a good opportunity to request data from the web server.
+-- Note that if the server is stopped then started again, this event will be raised again.
+-- @usage
+-- Event.add(Server.events.on_server_started,
+-- function()
+--      Server.try_get_all_data('regulars', callback)
+-- end)
+Public.events = {on_server_started = defines.events.on_server_started}
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server.lua]: wat?

File: server_commands.lua [hunk 4]
```diff
@@ -0,0 +1,28 @@
+local Poll = require 'features.gui.poll'
+local UserGroups = require 'features.user_groups'
+local Token = require 'utils.global_token'
+local Server = require 'server'
+
+--- This module is for the web server to call functions and raise events.
+-- Not intended to be called by scripts.
+-- Needs to be in the _G table so it can be accessed by the web server.
+ServerCommands = {}
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+ServerCommands.get_poll_result = Poll.send_poll_result_to_discord
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+ServerCommands.regular_sync = UserGroups.sync_regulars
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ServerCommands.donator_sync = UserGroups.sync_donators
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+function ServerCommands.raise_callback(func_token, data)
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    local func = Token.get(func_token)
+    func(data)
+end
+
+ServerCommands.raise_data_set = Server.raise_data_set
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ServerCommands.get_tracked_data_sets = Server.get_tracked_data_sets
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+function ServerCommands.server_started()
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: intended

File: server_commands.lua [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    script.raise_event(Server.events.on_server_started, {})
+end
+
+return ServerCommands
```
[the remark this answers is not in the document]
Comment (grilledham) [reply server_commands.lua]: wat?

**Comment by plague006** [comment]
In terms of file structure, can we move server and server_commands into either its own root folder `server` or `features/server` or something similar?

File: server.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (mheguy) [reply server.lua]: There is no factorio-defined defines.events.on_server_started and in luarcheckrc, `defines` is given as read-only, so it doesn't register you adding it on L30

File: server.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (mheguy) [reply server.lua]: In our luarcheckrc `defines` is given as read-only

[the remark this answers is not in the document]
Comment (Valansch) [reply server.lua]: Could we put this instead in Server.events.on_server_started?

File: server.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (Valansch) [reply server.lua]: stickler, sometimes you're an idiot.

**Review by Valansch — APPROVED** [review]
Love it. Now we can track everything!! :D

**Comment by Valansch** [comment]
You should maybe put this link into the file head of Server.lua: https://github.com/Refactorio/RedMew/pull/469
So we always find your code examples and explainations :P

File: server.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (mheguy) [reply server.lua]: Not really. Our luacheckrc says `defines` is read-only, so it doesn't register the extension of it.

**Comment by mheguy** [comment]
I actually think it's a good idea to include the link as documentation.

Re: Stickler. I think there are 3 ways to potentially proceed.
* We can do as Val suggested and put the new event into a different variable, not touching `defines`.
* We can define `defines` as a standard read-write global instead of read-only and add the `on_server_started` event to the definition. This *should* be as easy as just shuffling [L605[(https://github.com/Refactorio/RedMew/blob/b142939c5f1caa12b8372889ac92e58e2c53256b/.luacheckrc#L605) around a bit and adding `on_server_started` to the [defined fields](https://github.com/Refactorio/RedMew/blob/b142939c5f1caa12b8372889ac92e58e2c53256b/.luacheckrc#L609) of `defines`.
* We can ignore stickler, but at some point we'll likely want the entire repo to pass linting, so we would just be kicking this problem down the road.

Also just hats off for the amazing documentation inside and out.

**Comment by mheguy** [comment]
From a file management point of view, can we move the server-related files from the root directory? Either a folder named `server` or `features/server`?

**Comment by mheguy** [comment]
If this merges after #469, need to change server.lua Line 1 to require `utils.token`

**Comment by grilledham** [comment]
I'll move server and serverCommands into features, that makes sense.

I want the `on_server_started` event id to be in both `Server.events` and `defines.events`

There is a way to escape luacheck by adding comments to the code so that's another option to pass linting.

We can link this document at the start of server.lua, but I might also add the examples into `@usage` comments.

@plague006 I assume you mean #464 in reference to `utils.token`.

**Comment by mheguy** [comment]
GJ me linking to the PR we're inside of. :D

**Comment by grilledham** [comment]
It easier to add `defines.events.on_server_started` than remove it at a later date. So I've removed it for now. We can always add it back at a later date if we want to.

**Pull request merged**